### PR TITLE
Org config

### PR DESCRIPTION
**Pull request by maoo**
I am evaluating the cla-bot to validate Pull Requests across different projects that are hosted in our [github org](https://github.com/ColinEberhardt/cla-bot/issues/github.com/symphonyoss); for this reason, I need to be able to define a .clabot configuration at organisation level and avoid the project to overrule it.

The solution can be composed by 2 blocks:

If cla-bot fails resolving the project's .clabot file, it will fallback into https://github.com/<user/org_name>/clabot-config/.clabot
A configuration flag (ie forceOrgConfig) that forces .clabot to be resolved at organisation level
I am submitting a PR that I've [tested against a test repo](https://github.com/ssf-admin/clabot-test/pull/3#issuecomment-303710830); configuration is stored in [ssf-admin/clabot-config](https://github.com/ssf-admin/clabot-config).

Please note; the PR is not ready for merge; documentation and testing is missing; feedback is welcome, especially considering my entry-level node skills.

### CODE REVIEW

**Comment by ColinEberhardt** [comment]
Thanks @maoo - looks like a very good idea to me. A few thoughts ...

Conceptually, this seems very similar to the way eslint allows you to centralise configuration, e.g.: https://github.com/ColinEberhardt/cla-bot/blob/master/.eslintrc.json#L3

As an alternative, how about having an optional `extends` property within `.clabot` that identifies a GitHub project via the slug, e.g.

```
{
  "extends": "symphonyoss/clabot-config"
}
```

If `extends` exists, it would fetch `symphonyoss/clabot-config/.clabot`, then merge any local configuration (followed by the defaults)

What do you think?

**Comment by maoo** [comment]
I like the idea of the `extends` property, but it poses two additional challenges, for the use case of the Foundation:
1. Every project must have a `.clabot` file ; this _could_ be solved by adding a [contribution (legal) requirement](https://symphonyoss.atlassian.net/wiki/display/FM/Legal+Requirements), but it adds one step in the contribution process
2. It allows projects to overrule the contributors whitelist (either embedding the list in `.clabot` or defining a `contributorListUrl` property) and grant access to someone unauthorized; I'd like to have a switch to enable/disable this possibility; maybe we could use a [protected branch](https://help.github.com/articles/about-protected-branches/) in github, where we restrict access to `.clabot` ?

**Comment by ColinEberhardt** [comment]
> Every project must have a .clabot file ; this could be solved by adding a contribution (legal) requirement, but it adds one step in the contribution process

I know this sounds like a pain, but it's not uncommon. Virtually every tool (travis, appveyor, eslint, typescript, babel, rollup ...) requires some form of configuration file. The advantage of having this file is that it is explicit - i.e. it is obvious to someone who inspects the repo that this is the configuration file for the robot!

> It allows projects to overrule the contributors whitelist (either embedding the list in .clabot or defining a contributorListUrl property) and grant access to someone unauthorized.

That's true, although anything that depends on a local config file would have the same problem. With your current implementation someone could just add a `.clabot` file, removing the org-level fallback.

Project maintainers can do all kinds of damage already, change LICENCEs, include copyright code, etc ... I think sneaking in a contributor is probably quite low risk!

**Comment by pmonks** [comment]
@ColinEberhardt I think @maoo's proposal is to add a .clabot file that's managed by the operator of the cla-bot instance and therefore inaccessible to project teams.  Then the only thing necessary is to add a configuration option to cla-bot (as @maoo describes above) that forces cla-bot to only look at that file (and completely ignore any .clabot files residing in the project itself).

This is a requirement for organisations (such as ours) where CLAs are managed centrally, outside the project teams.

**Comment by ColinEberhardt** [comment]
Hi @pmonks thanks for the input - I can see your point. 

With cla-bot I am modelling the way it works on services like travis and appveyor, which is why I have created a 'hosted' version which you add as a [GitHub App](https://github.com/apps/cla-bot), with configuration stored locally within the repo. 

Although, I guess for Symphony you want to self-host so that you are not dependent on someone else hosting it for you? In this instance, I think that it is better to use an environment variable that causes the bot to use a `.clabot` file from some other source, rather than the local repo version.

How about adding an optional `CLABOT_CONFIG_URL` environment variable, that in your case points to https://github.com/ssf-admin/clabot-config/blob/master/.clabot ? If this environment variable is present, any project local `.clabot` files would be ignored. And if it is not present, the bot follows the current behaviour.

**Comment by pmonks** [comment]
Actually we'd be more than happy to have someone else host the bot - that's not a concern at all.  The issue is that the Symphony Software Foundation's CLA model (which is the same as most other open source Foundations, fwiw) is tripartite rather than the more widely known bipartite model.  By this I mean that contributors (party 1) submit contributions to Foundation-hosted projects (party 2), but license those contributions to the Foundation (party 3).  For several reasons the Foundation (party 3) cannot delegate CLA compliance validation to projects (party 2), which is effectively what cla-bot's current model does (by having the master `.clabot` file managed by the project as part of their source code).

And just to be clear, we're not proposing to take away support for the bipartite model, since it is widely used in other contexts.  @maoo's proposal instead suggests adding configurable support for the tripartite model (albeit via a runtime configuration switch that would likely force us to run a separate instance of cla-bot - there may be better ways to design this so that we can have our someone-else-hosts-the-bot cake and eat the tripartite-model-support cake too).

**Comment by ColinEberhardt** [comment]
Thanks for the detailed explanation, that was really useful.

>  ( ... there may be better ways to design this so that we can have our someone-else-hosts-the-bot cake and eat the tripartite-model-support cake too)

I'm keen to investigate that further. I'll have a think!

**Comment by pmonks** [comment]
Cool!  Yeah one half-baked thought I had was that cla-bot could check whether the project being contributed to is owned by a GitHub user vs a GitHub organisation, and then either use the bipartite or tripartite model based on that (pulling the `.clabot` file from the organisation somehow - how that would be done is TBD).

It's assuming quite a bit (i.e. that all organisation projects use a tripartite model), but that split is one possible way for a single runtime instance of the bot to be able to figure this out, based on input data already available to it.

wdyt?

**Comment by ColinEberhardt** [comment]
I really think it's quite simple, and doesn't need to depend on whether a repo is owned by an individual or organisation ...

Given a webhook indicating a pull request for project `github.com/foo/bar`:

1. The bot first tries to obtain the configuration from `github.com/foo/clabot-config/.clabot`.
2. If the file doesn't exist, it tries to obtain the configuration from `github.com/foo/bar/.clabot`.

In other words, to set configuration at an individual user or organisation level the bot requires that you create a `clabot-config` project and add the `.clabot` file there. You can then of course ensure that you restrict access to this repo (or even make it private).

One minor problem here is that when using the bot as GitHub App, its webhook is invoked for every repo within an organisation. In your current deployment you are manually adding webhooks on a per-project basis. Would you really want the bot to verify commits on every repo? Perhaps `.clabot` should also support a white-list of projects to check CLAs against. Anyhow, I think this problem can be tackled later on.

**Comment by pmonks** [comment]
Yeah I like that approach, though I don't think the `clabot-config` project can be private, since then clabot itself won't be able to access it.  Access to `github.com/foo/bar` (which clabot will have, by virtue of running in a webhook) doesn't imply access to `github.com/foo/clabot-config`, at least as far as I understand GitHub's permissioning model.

**Comment by ColinEberhardt** [comment]
cla-bot isn't an anonymous user, instead it uses a personal access token (when installed as a webhook), or an installation access token (when installed as an app). In both cases you can request privileged access ...

screen shot 2017-05-24 at 21 57 59

So access to private repos is entirely possible :-)

**Comment by maoo** [comment]
> I really think it's quite simple, and doesn't need to depend on whether a repo is owned by an individual or organisation ...
> 
> Given a webhook indicating a pull request for project github.com/foo/bar:
> 
> 1. The bot first tries to obtain the configuration from github.com/foo/clabot-config/.clabot.
> 2. If the file doesn't exist, it tries to obtain the configuration from github.com/foo/bar/.clabot.
> 
> In other words, to set configuration at an individual user or organisation level the bot requires that you create a clabot-config project and add the .clabot file there. You can then of course ensure that you restrict access to this repo (or even make it private).

@ColinEberhardt - I've [changed the logic](https://github.com/ColinEberhardt/cla-bot/pull/11/commits/30885b27007e7cfcdefb1b6da842892583ed989d#diff-168726dbe96b3ce427e7fedce31bb0bcR31) to check github.com/foo/clabot-config/.clabot first, then fallback to github.com/foo/bar/.clabot , as you documented above (previously it was the opposite)

> Yeah I like that approach, though I don't think the clabot-config project can be private, since then clabot itself won't be able to access it. Access to github.com/foo/bar (which clabot will have, by virtue of running in a webhook) doesn't imply access to github.com/foo/clabot-config, at least as far as I understand GitHub's permissioning model.

I turned `clabot-config` project to private (to validate @pmonks last comment) and realised that the `contributorListUrl` was fetched without using the `GITHUB_ACCESS_TOKEN`; I added a `contributorListGithubUrl` configuration property, which now allows to fetch an `https://api.github.com/...` URL with the token attached to it (using `githubRequest` instead of `requestp` function), which is why I had to move some functions around.

**Comment by ColinEberhardt** [comment]
Thanks for the update @maoo - definitely moving in the right direction.

**Comment by ColinEberhardt** [comment]
Just had a quick look at the code - the restructuring / moving around of `githubRequest` etc ... looks fine to me 👍

**Comment by ColinEberhardt** [comment]
Now that #13 has been merged, it's probably worth rebasing so that travis works for this PR.

You should start to see linting errors instead :-)

**Comment by maoo** [comment]
@ColinEberhardt - Rebasing done.

Linting passes (after a quick fix), but I'm afraid you'd need to merge https://github.com/ColinEberhardt/cla-bot/pull/14 and I'd need to rebase again, in order to get a green build.

**Comment by ColinEberhardt** [comment]
see #15 - Travis now has a green tick on PR branches :-)

File: contributionVerifier.js [hunk 1]
```diff
@@ -1,9 +1,22 @@
 const requestp = require('./requestAsPromise');
+const {githubRequest,getReadmeContents} = require('./githubApi');
+
+const clabotToken = process.env.GITHUB_ACCESS_TOKEN;
 
 const contributorArrayVerifier = (contributors) =>
   (committers) =>
     Promise.resolve(committers.filter(c => contributors.indexOf(c) === -1));
 
+const configFileFromGithubUrlVerifier = (contributorListGithubUrl) =>
```
Comment (ColinEberhardt) [inline contributionVerifier.js:10]: No need for braces ..,

```
const configFileFromGithubUrlVerifier = (contributorListGithubUrl) =>
  (committers) => githubRequest({
      url: contributorListGithubUrl,
      method: 'GET'
    },clabotToken)
    .then((body) => githubRequest(getReadmeContents(body),clabotToken))
    .then((contributors) => contributorArrayVerifier(contributors)(committers));
```

File: default.json [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
   "message": "Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don\"t seem to have you on file. In order for us to review and merge your code, please contact the project maintainers to get yourself added.",
-  "label": "cla-signed"
+  "label": "cla-signed",
+  "forceOrgConfig" : false
```
Comment (ColinEberhardt) [inline default.json]: We're not using this any more.

File: githubApi.js [hunk 3]
```diff
@@ -1,3 +1,29 @@
+const requestp = require('./requestAsPromise');
+
+const getOrgConfigUrl = (repositoryUrl) => {
+  var gh = require('parse-github-url');
```
Comment (ColinEberhardt) [inline githubApi.js]: can you use `const` instead? I should update the lint rules to enforce this!

File: githubApi.js [hunk 4]
```diff
@@ -1,3 +1,29 @@
...
+  ghData = gh(repositoryUrl);
+  ghUrl = "https://" + ghData.host + "/repos/" + ghData.owner + "/clabot-config/contents/.clabot";
+  console.log("clabot org URL - "+ghUrl);
```
Comment (ColinEberhardt) [inline githubApi.js]: This isn't necessary, the URLs that the bot requests are logged already

File: index.js [hunk 5]
```diff
@@ -26,20 +26,19 @@ exports.handler = ({ body }, lambdaContext, callback) => {
...
   console.log(`Checking CLAs for PR ${context.webhook.pull_request.url}`);
 
-  githubRequest(getReadmeUrl(context))
-    .then(body => githubRequest(getReadmeContents(body)))
+  githubRequest(getOrgConfig(context),clabotToken)
+    .then(body => {
+      if (!body.name || body.name == "") {
+        console.log("Couldn't fetch .clabot from Github organisation project; trying at project level");
+        return githubRequest(getReadmeUrl(context),clabotToken);
+      } else {
```
Comment (ColinEberhardt) [inline index.js]: This `else` block isn't needed

**Review by ColinEberhardt — CHANGES_REQUESTED** [review]

**Comment by maoo** [comment]
Shall we close this and keep on working on #16 ?

**Comment by ColinEberhardt** [comment]
merged via #16

**Pull request closed without merging**